### PR TITLE
feat: make display limit configurable from user settings

### PR DESCRIPTION
**Pull request by namit-chandwani**
## Linked Issues
- Closes https://github.com/sourcegraph/sourcegraph/issues/36691

## Motivation and Context:
 
- To provide users the flexibility to modify the display limit as per their requirement
- However, it's worth noting that increasing the display limit could potentially result in decreased performance

## Changes Made:
 
- On the frontend:
	- Added a new field named `search.displayLimit` to the User settings
	- Started using the `search.displayLimit` value while performing stream search
- On the backend:
	- No changes

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (altering code without changing its external behaviour)
- [ ] Documentation change
- [ ] Other

## Checklist:

- [x] Development completed
- [x] Comments added to code (wherever necessary)
- [x] Documentation updated (if applicable)
- [x] Tested changes locally

## Follow-up tasks (if any):

- [ ] Update the documentation (PR link: https://github.com/sourcegraph/docs/pull/279)

## Test Plan

 - Had setup the codebase locally and tested the entire flow locally end-to-end
 - Screen recording which shows the display limit from the newly added site configuration parameter being used while performing code search:
 
 https://github.com/sourcegraph/sourcegraph/assets/51479159/a7807447-7a9e-4710-a8b9-564b11db2e1d
    
## Additional Comments
- Please let me know if any further changes are required elsewhere

### CODE REVIEW

**Comment by stefanhengl** [comment]
Thanks for the PR! I really appreciate the thorough description and the video! Nice job! However, I think we should avoid changing the API if we can. So another option would be to keep the query parameter, but instead let the client get the display limit from the config. WDYT?

Thinking about it ... I know this was my suggestion, but I think "settings" might be a better choice than site config, because users can override it instead of having to ask an admin to change it for everyone. We could use "search.displayLimit" as key.

There is a function `useSettings()` in "settings/settings.tsx" which might come in handy.

This design wouldn't require any backend or API changes.

**Comment by namit-chandwani** [comment]
@stefanhengl Hey, I was travelling back then so didn't get a chance to look into this.

I'll resume my work on this feature today and push the new changes within a few days :)

File: client/shared/src/search/stream.ts [hunk 1]
```diff
@@ -500,15 +500,6 @@ export interface StreamSearchOptions {
...
     zoektSearchOptions?: string
 
-    /**
-     * Limits the number of matches sent down. Note: this is different to the
-     * count: in the query. The search will continue once we hit displayLimit
-     * and updated filters and statistics will continue to stream down.
-     *
-     * If unset all results are streamed down.
-     */
-    displayLimit?: number
```
[the remark this answers is not in the document]
Comment (namit-chandwani) [reply client/shared/src/search/stream.ts:510]: Sure, I'll give it a try

**Comment by namit-chandwani** [comment]
Hey @stefanhengl, apologies for the delay.

I've pushed the changes based on the updated requirements now (ie. no change on the backend API and displayLimit needs to be configured on the user settings instead of the site admin config).

Here's the new screen recording of the testing performed locally after making these changes:


https://github.com/sourcegraph/sourcegraph/assets/51479159/53887788-4ef3-4cf7-ab7a-a57a51cb64e5

Request you to review the changes whenever free

- New documentation PR for these changes: https://github.com/sourcegraph/docs/pull/279

**Review by stefanhengl — APPROVED** [review]
LGTM! This turned out really nice! Thanks for your work!

File: client/web-sveltekit/src/routes/search/+page.ts [hunk 2]
```diff
@@ -116,6 +116,7 @@ export const load: PageLoad = ({ url, depends }) => {
             featureOverrides: [],
             chunkMatches: true,
             searchMode,
+            // TODO: populate this from user settings
```
Comment (jtibshirani) [inline client/web-sveltekit/src/routes/search/+page.ts:119]: CC @sourcegraph/code-search I added this TODO. Our external contributor @namit-chandwani has been very patient with us, and I didn't want to hold up this PR to implement it.

Comment (camdencheek) [reply client/web-sveltekit/src/routes/search/+page.ts:119]: Thanks! Will follow up after it's merged

**Comment by jtibshirani** [comment]
Merging on behalf of @stefanhengl since he's out. Thanks again for your contribution !

File: client/web-sveltekit/src/routes/search/+page.ts (hunk 2 above)
Comment (camdencheek) [reply client/web-sveltekit/src/routes/search/+page.ts:119]: [Closing the loop](https://github.com/sourcegraph/sourcegraph/pull/62681)

**Pull request merged**